btb/hyper_parameter.py

```diff
@@ -21,37 +21,64 @@ class ParamTypes(object):
 
 # HyperParameter object
 class HyperParameter(object):
-    def __new__(cls, typ, rang):
-        if cls is HyperParameter:
-            return super(HyperParameter, cls).__new__(CLASS_GENERATOR[typ])
-        else:
-            return super(HyperParameter, cls).__new__(cls)
-
-    def __init__(self, rang, cast):
-        for i, val in enumerate(rang):
-            if val is None:
-                # the value None is allowed for every parameter type
-                continue
-            rang[i] = cast(val)
+
+    param_type = None
+    is_integer = False
+
+    _subclasses = []
+
+    @classmethod
+    def _get_subclasses(cls):
+        subclasses = []
+        for subclass in cls.__subclasses__():
+            subclasses.append(subclass)
+            subclasses.extend(subclass._get_subclasses())
+
+        return subclasses
+
+    @classmethod
+    def subclasses(cls):
+        if not cls._subclasses:
+            cls._subclasses = cls._get_subclasses()
+
+        return cls._subclasses
+
+    def __new__(cls, param_type=None, rang=None):
+        if param_type:
+            for subclass in cls.subclasses():
+                if subclass.param_type == param_type:
```

Comment:
No, they aren't. The ParamType class has attributes that work as helpers, but we can actually work with any string, as far as they have the right value (f.e. `"int"`).
And it is not recommended to check string values using the `is` operator because it depends on the Python implementation.